### ISSUE TITLE
Question about YOLO11 train on custom dataset（ImageNet VID）

### ISSUE THREAD

**Issue by xiaokj37** [ultralytics/ultralytics#21687]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

When training on a custom dataset, the validation process does not execute at periodic intervals. My initial attempts using the val_period command-line argument did not trigger validation as expected.

To have more explicit control, I then tried using a Python script with a custom callback (on_train_epoch_end) to programmatically enable validation every 5 epochs. Despite this, the behavior remains the same: the console output only shows training progress, and no validation steps are run intermittently.

Consequently, the results.csv file shows that while training metrics are updated correctly with each epoch, all validation metrics (val/box_loss, metrics/mAP50-95(B), etc.) remain static after the initial validation run.

Here is the setup that demonstrates the issue using a Python script and callbacks.

1. Python Training Script:
This script uses a callback to set trainer.args.val = True every 5th epoch.

from ultralytics import YOLO

def validate_interval(trainer):
    """Callback to enable validation every 5 epochs."""
    # The validation should run when the epoch number (1-indexed) is a multiple of 5
    if (trainer.epoch + 1) % 5 == 0:
        trainer.args.val = True
    else:
        trainer.args.val = False

def train_yolo_model():
    # Load a model
    model = YOLO("yolo11s.pt")

    # Add the callback
    model.add_callback("on_train_epoch_end", validate_interval)

    # Start training
    # Note: val is not set to True here, as the callback should control it.
    model.train(
        data="/root/share/datasets/imagenet_vid/imagenet_vid_yolo_format_small/imagenet_vid.yaml",
        epochs=500,
        imgsz=640,
        batch=256,
        workers=8,
        device=[0,1,2,3],
    )

if __name__ == "__main__":
    train_yolo_model()

2. Dataset Configuration (imagenet_vid.yaml):

path: /root/share/datasets/imagenet_vid/imagenet_vid_yolo_format_small
train: images/train
val: images/val

# Number of classes
nc: 30

# Class names
names:
- airplane
- antelope
- bear
- bicycle
- bird
- bus
- car
- cattle
- dog
- domestic_cat
- elephant
- fox
- giant_panda
- hamster
- horse
- lion
- lizard
- monkey
- motorcycle
- rabbit
- red_panda
- sheep
- snake
- squirrel
- tiger
- train
- turtle
- watercraft
- whale
- zebra

3. results.csv Observation:
The resulting CSV file shows that validation metrics do not update.

epoch,time,train/box_loss,train/cls_loss,train/dfl_loss,metrics/precision(B),metrics/recall(B),metrics/mAP50(B),metrics/mAP50-95(B),val/box_loss,val/cls_loss,val/dfl_loss,lr/pg0,lr/pg1,lr/pg2
1,190.587,0.99882,1.89765,1.18934,0.60665,0.51755,0.56409,0.40161,1.00093,1.62113,1.14776,0.00332618,0.00332618,0.00332618
2,333.565,0.92117,0.88398,1.11925,0.60665,0.51755,0.56409,0.40161,1.00093,1.62113,1.14776,0.00664633,0.00664633,0.00664633
3,479.105,0.98305,0.91889,1.15317,0.60665,0.51755,0.56409,0.40161,1.00093,1.62113,1.14776,0.00995328,0.00995328,0.00995328
4,623.217,1.01605,0.95448,1.17702,0.60665,0.51755,0.56409,0.40161,1.00093,1.62113,1.14776,0.0099406,0.0099406,0.0099406
5,766.532,0.96945,0.83322,1.14661,0.60665,0.51755,0.56409,0.40161,1.00093,1.62113,1.14776,0.0099208,0.0099208,0.0099208
6,908.925,0.93829,0.7684,1.12858,0.60665,0.51755,0.56409,0.40161,1.00093,1.62113,1.14776,0.009901,0.009901,0.009901

Expected behavior
I expect the callback to successfully trigger the validation process at the end of epochs 5, 10, 15, etc. The console should show the validation progress, and the results.csv file should contain updated validation metrics for those specific epochs.

Environment
I have confirmed I am using a very recent version of the ultralytics package.

Ultralytics 8.3.178 🚀 Python-3.10.12 torch-2.8.0+cu128 CUDA:0 (NVIDIA GeForce RTX 4090, 24092MiB) Setup complete ✅ (48 CPUs, 125.5 GB RAM, 6703.0/14783.5 GB disk)  OS                  Linux-6.8.0-65-generic-x86_64-with-glibc2.35 Environment         Docker Python              3.10.12 Install             pip RAM                 125.50 GB CPU                 Intel Xeon Gold 5317 3.00GHz CUDA                12.8  torch               ✅ 2.8.0>=1.8.0 torchvision         ✅ 0.23.0>=0.9.0 ... (other packages are up to date) 

Thank you for your help in looking into this.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @xiaokj37, thank you for your interest in Ultralytics 🚀! This is an automated response to help you get moving quickly—an Ultralytics engineer will also assist you here soon.

We recommend a visit to the Docs for new users where you can find many Python and CLI usage examples and where many of the most common questions may already be answered:
- Docs: https://docs.ultralytics.com/
- Python: https://docs.ultralytics.com/usage/python/
- CLI: https://docs.ultralytics.com/usage/cli/

If this is a 🐛 Bug Report, please provide a minimum reproducible example to help us debug it:
- MRE: https://docs.ultralytics.com/help/minimum-reproducible-example/

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our Tips for Best Training Results:
- Training Tips: https://docs.ultralytics.com/guides/model-training-tips/

Join the Ultralytics community where it suits you best:
- Discord (real-time chat) 🎧: https://discord.com/invite/ultralytics
- Discourse (in-depth Q&A): https://community.ultralytics.com/
- Subreddit (community threads): https://www.reddit.com/r/Ultralytics/

## Upgrade

Please upgrade to the latest ultralytics package including all requirements in a Python>=3.8 environment with PyTorch>=1.8 to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

- Requirements: https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml
- Python: https://www.python.org/
- PyTorch: https://pytorch.org/get-started/locally/

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including CUDA/CUDNN, Python and PyTorch preinstalled):

- Notebooks with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- Google Cloud Deep Learning VM. See GCP Quickstart Guide: https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/
- Amazon Deep Learning AMI. See AWS Quickstart Guide: https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/
- Docker Image. See Docker Quickstart Guide: https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/ <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all Ultralytics CI tests are currently passing. CI tests verify correct operation of all YOLO Modes and Tasks on macOS, Windows, and Ubuntu every 24 hours and on every commit:
- Modes: https://docs.ultralytics.com/modes/
- Tasks: https://docs.ultralytics.com/tasks/

To help us investigate your report on YOLO11 validation intervals:
1. Please share a minimal reproducible example that runs in under a few minutes and reproduces the missing periodic validation. This could be a tiny subset of ImageNet VID or a synthetic dataset with the same behavior.
2. Include exact code, CLI args used (if any), your callbacks setup, and the full terminal log from start to at least 2 expected validation points.
3. Attach your results.csv and the corresponding run directory structure.
4. Confirm the exact ultralytics version via:
   - python -c "import ultralytics, sys; print(ultralytics.__version__, sys.version)"
5. Try reproducing with the latest version after running pip install -U ultralytics and report if the behavior changes.

Thanks for the detailed description and environment info—once we have an MRE, we can dig in faster 🧪

**Comment by Y-T-G** [comment]
You can try this:

https://github.com/ultralytics/ultralytics/issues/19575#issuecomment-2708201247

**Comment by xiaokj37** [comment]
To have more explicit control, I then tried using a Python script with a custom callback (on_train_epoch_end) to programmatically enable validation every 5 epochs. However, this callback approach (model.add_callback) does not appear to function correctly in a multi-GPU training environment. Despite this setup, the behavior remains the same: the console output only shows training progress, and no validation steps are run intermittently as defined.

Consequently, the results.csv file shows that while training metrics are updated correctly with each epoch, all validation metrics (val/box_loss, metrics/mAP50-95(B), etc.) remain static after the initial validation run.
Furthermore, I've noticed a confusing behavior where validation seems to run approximately every 20 epochs, which contradicts both my callback logic (every 5 epochs) and the default behavior. This adds to the uncertainty of how validation is being controlled.

**Comment by Y-T-G** [comment]
Custom callbacks don't work for multi-GPU training because it's launched in a different process
